doc/reference/trd-unrestricted-allow-syscalls.md

```diff
@@ -0,0 +1,150 @@
+able to read a buffer without first revoking it from the kernel with a
+Read-Write Allow. These cases are situations when the cost of a
+Read-Write Allow system call is an unacceptable overhead for
+accessing the data.
+
+This allows capsules that support the unrestricted allow call to communicate with
+applications without buffers needing to be re-allowed. For example a capsule
+might want to report statistics to a userspace app. It could do this by letting
+the app perform a unrestricted allow call to allocate a buffer. Then the capsule can
+write statistics to the buffer and at any time the app can read the statistics
+from the buffer.
+
+2 System Call API
+=================================
+
+2.1 Unrestricted Allow (Class ID: 7)
+---------------------------------
+The unrestricted allow syscall follows the same expectations and requirements
+as described for the Read-Write syscall in section TRD104 4.4, with the
+exceptionthat apps are explicity allowed to read/write buffers that have
+been passed to the kernel.
+
+The register arguments for Read-Write Allow system calls are as
+follows. The registers r0-r3 correspond to r0-r3 on CortexM and a0-a3
+on RISC-V.
+
+| Argument         | Register |
+|------------------|----------|
+| Driver number    | r0       |
+| Buffer number    | r1       |
+| Address          | r2       |
+| Size             | r3       |
+
+The Tock kernel MUST check that the passed buffer is contained within
+the calling process's writeable address space. Every byte of a passed
+buffer must be readable and writeable by the process. Zero-length
+buffers may therefore have arbitrary addresses. If the passed buffer is
+not complete within the calling process's writeable address space, the
+kernel MUST return a failure result with an error code of `INVALID`.
+The buffer number specifies which buffer this is. A driver may
+support multiple allowed buffers.
+
+The return variants for Read-Write Allow system calls are `Failure
+with 2 u32` and `Success with 2 u32`.  In both cases, `Argument 0`
+contains an address and `Argument 1` contains a length. When a driver
+implementing the Read-Write Allow system call returns a failure
+result, it MUST return the same address and length as those that were passed
+in the call. When a driver implementing the Read-Write Allow system call
+returns a success result, the returned address and length MUST be those
+that were passed in the previous call, unless this is the first call.
+On the first successful invocation of a particular Read-Write Allow system
+call, an driver implementation MUST return address 0 and size 0.
+
+The syscall class ID is shown below:
+
+| Syscall Class    | Syscall Class Number |
+|------------------|----------------------|
+| Read-Write Allow |           7          |
+
+The standard access model for unrestricted allowed buffers is that userspace can read
+and write to a buffer while the kernel can also read or write. Syncronisation
```

Comment:
My recollection of this system call and the discussion around it was that it was for kernel-write/userspace-read use cases. I am against expanding it to allow arbitrary access then just punting it to the system call implementer to make sure it's OK. As I think the earlier discussion (e.g., #2381 ) showed, this is a notoriously difficult problem that's easy to get wrong. 

The consistency mechanisms you use for a userspace-read/kernel-write buffer are different than for kernel-read/userspace-write one, because of Tock's concurrency model. E.g., monotonic counters don't work for the latter one unless you have a timeout/retry mechanism, and in that case you have no assurance of forward progress (you could always hit it mid-write).